```{r}
# MAKE SURE THE WD IS THE ROOT PROJECT FOLDER

# Get the data
data.clinical <- read.delim("data/data_clinical_patient.txt", sep = "\t", header = TRUE, comment.char = "#")
data.mutation <- read.delim("data/data_mutations.txt", sep = "\t", header = TRUE, comment.char = "#")
data.expression <- read.delim("data/RNAseq_BRCA.csv", sep = ",", header = TRUE, comment.char = "#")
```

```{r}
# Build the patient-mutation matrix
mutation.patients <- unique(substr(data.mutation$Tumor_Sample_Barcode, 1, 12))
clinical.patients <- data.clinical$PATIENT_ID
expression.patients <- gsub("\\.", "-", substr(colnames(data.expression), 1, 12)[-1])

# Get the patients that we have full data for
unique.patients.full.data <- Reduce(intersect, list(mutation.patients, clinical.patients, expression.patients))

# Make a new column with the cleaned patient ID (same as other datasets)
data.mutation$Tumor_Sample_Barcode_Cleaned <- substr(data.mutation$Tumor_Sample_Barcode, 1, 12)

# Non-coding transcript exon variant?
important_mutations <- data.mutation[which(data.mutation$IMPACT %in% c("HIGH", "MODERATE")), ]

# Get the important mutations for full-data patients
important_mutations_full_data <- important_mutations[which(important_mutations$Tumor_Sample_Barcode_Cleaned %in% unique.patients.full.data), ]

# Make feature matrix
feature_mat <- table(important_mutations_full_data$Tumor_Sample_Barcode_Cleaned, important_mutations_full_data$Hugo_Symbol)

# Turn into binary matrix
feature_mat[feature_mat > 1] <- 1
```

Clustering

Filter for the top ~20 mutated genes
```{r}
quantile(colSums(feature_mat), probs = 0.999)
```
Therefore, any genes with 51 or more mutations are in the 99.8th percentile.

```{r}
feature_mat_filtered <- feature_mat[, colSums(feature_mat) >= 51]
str(feature_mat_filtered)
```
The feature matrix before filtering had 16503 genes. This matrix has only the most mutated genes (>51 patients) and only has 17 genes.

```{r message = FALSE}
library(tidyverse)
feature_df_filtered <- as.data.frame(feature_mat_filtered)
feature_df_filtered <- pivot_wider(feature_df_filtered, names_from = Var2, values_from = Freq)
# Save the patients
patient.order <- feature_df_filtered$Var1
# Remove the patients column
feature_df_filtered <- feature_df_filtered[, -1]
```

Visualize most mutated genes
```{r}
library(ggplot2)

# Create frequency table
freq <- colSums(feature_df_filtered == 1)

# Convert frequency table to a data frame
data <- data.frame(column = names(freq), frequency = freq)

# Sort the data frame by frequency in descending order
data <- data[order(data$frequency, decreasing = TRUE), ]

# Create a bar plot using ggplot
ggplot(data, aes(x = column, y = frequency)) +
  geom_col() + theme(axis.text.x = element_text(angle = 45,hjust=1)) + 
  scale_x_discrete(limits = data$column) +
  labs(x = "Mutated Genes", y = "Frequency (# of patients)")
```

PCA:
```{r warning = FALSE}
feature.pca <- prcomp(feature_df_filtered, center = TRUE)

library(ggbiplot)
ggbiplot(feature.pca, var.axes = FALSE, ellipse = TRUE)
summary(feature.pca)
```

Hierarchical clustering:

We performed hierarchical clustering on the filtered data using single linkage, average linkage, complete linkage and Ward linkage methods. The Ward method yielded the most evenly distributed clusters.
```{r}
set.seed(123)
dist_mat <- dist(feature_df_filtered, method = 'binary')
hclust_ward <- hclust(dist_mat, method = 'ward.D')
```
```{r}
cut_ward <- cutree(hclust_ward, h = 40)
plot(hclust_ward, main = "Hierarchical Clustering Dendogram")
rect.hclust(hclust_ward, h = 40, border = 2:6)
abline(h = 40, col = 'red')
```
```{r}
cluster.size.hc <- table(cut_ward)
cluster.size.hc
```

Next, we performed hierarchical clustering on the PCA data.
```{r}
set.seed(123)

#Use top 12 pcs to perserve ~85% of the variance
dist_mat_pca <- dist(feature.pca$x[,1:12], method = 'euclidean')
hclust_pca <- hclust(dist_mat_pca, method = 'ward.D')
```
```{r}
cut_pca <- cutree(hclust_pca, k=5)
plot(hclust_pca, main = "PCA Hierarchical Clustering Dendogram")
rect.hclust(hclust_pca,k=5, border = 2:6)
#abline(h=80, col = 'red')
```
```{r}
cluster.size.hc.pca <- table(cut_pca)
cluster.size.hc.pca
```
