---
title: "MutationAnalysis-Final"
author: "Aadesh Mehra"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r}
library(readr)
clinicalData <- read.table("C://Users//Aadesh//OneDrive - UBC//Desktop//bmeg - third//BMEG310 Project//BMEG310Project//TCGA-BRCA//data_clinical_patient.txt", sep="\t", header=TRUE)

MutationData <- read.delim("C://Users//Aadesh//OneDrive - UBC//Desktop//bmeg - third//BMEG310 Project//BMEG310Project//TCGA-BRCA//data_mutations.txt")

RNAseqData <- read.table("C://Users//Aadesh//OneDrive - UBC//Desktop//bmeg - third//BMEG310 Project//BMEG310Project//TCGA-BRCA//RNAseq_BRCA.csv", sep=",", header=TRUE)
```


```{r}

mutation.patients <- unique(substr(MutationData$Tumor_Sample_Barcode, 1, 12))
clinical.patients <- clinicalData$PATIENT_ID
expression.patients <- gsub("\\.", "-", substr(colnames(RNAseqData), 1, 12)[-1])

unique.patients.full.data <- Reduce(intersect, list(mutation.patients, clinical.patients, expression.patients))

MutationData$Tumor_Sample_Barcode_Cleaned <- substr(MutationData$Tumor_Sample_Barcode, 1, 12)



```

```{r}

# Non-coding transcript exon variant?
important_mutations <- MutationData[which(MutationData$IMPACT %in% c("HIGH", "MODERATE")), ]

# Get the important mutations for full-data patients
important_mutations_full_data <- important_mutations[which(important_mutations$Tumor_Sample_Barcode_Cleaned %in% unique.patients.full.data), ]

# Make feature matrix
feature_mat <- table(important_mutations_full_data$Tumor_Sample_Barcode_Cleaned, important_mutations_full_data$Hugo_Symbol)

# Turn into binary matrix
feature_mat[feature_mat > 1] <- 1

```

```{r}
#Filter for the top ~20 mutated genes
quantile(colSums(feature_mat), probs = 0.9995)
#Therefore any genes with 28 or more mutations are in the 99.9th percentile.

```

```{r}
#Only analyze genes with greater than 51 mutations to filter down data
feature_mat_filtered <- feature_mat[, colSums(feature_mat) >= 43]
print(ncol(feature_mat_filtered))
```
The feature matrix before filtering had 16,761 genes. This matrix has only the most mutated genes (>=51 patients) and only has 17 genes.

```{r}
#Convert to dataframe matrix form
library(tidyverse)
feature_df_filtered <- as.data.frame(feature_mat_filtered)
feature_df_filtered <- pivot_wider(feature_df_filtered, names_from = Var2, values_from = Freq)
# Save the patients
patient.order <- feature_df_filtered$Var1
# Remove the patients column
feature_df_filtered <- feature_df_filtered[, -1]
pheatmap_mat <- t(feature_df_filtered)

```


Visualize most mutated genes
```{r}
library(ggplot2)

# Create frequency table
freq <- colSums(feature_df_filtered == 1)

# Convert frequency table to a data frame
data <- data.frame(column = names(freq), frequency = freq)

# Sort the data frame by frequency in descending order
data <- data[order(data$frequency, decreasing = TRUE), ]

# Create a bar plot using ggplot
ggplot(data, aes(x = column, y = frequency)) +
  geom_col() + theme(axis.text.x = element_text(angle = 45,hjust=1)) + 
  scale_x_discrete(limits = data$column) +
  labs(x = "Mutated Genes", y = "Frequency (# of patients)")
```

```{r}
#Clustering using Ward method without PCA
set.seed(123)
dist_mat <- dist(feature_df_filtered, method = 'euclidean')
hclust_ward <- hclust(dist_mat, method = 'ward.D')
```

```{r}
cut_ward <- cutree(hclust_ward, k = 2)
plot(hclust_ward, main = "Hierarchical Clustering Dendogram")
rect.hclust(hclust_ward, k = 2, border = 2:6)
abline(h = 40, col = 'red')
cut_ward
```
```{r}
library(pheatmap)
reduce.mat <- pheatmap_mat
res <- pheatmap(reduce.mat,
         cluster_rows = F,
         show_colnames=FALSE)
```

```{r}
#PCA clustering:

feature.pca <- prcomp(feature_df_filtered, center = TRUE)

library(ggbiplot)
ggbiplot(feature.pca, var.axes = FALSE, ellipse = TRUE)
summary(feature.pca)
```

```{r}
#Next, we performed hierarchical clustering on the PCA data for 2 clusters.

set.seed(123)

#Use top 12 pcs to perserve ~85% of the variance
dist_mat_pca <- dist(feature.pca$x[,1:15], method = 'euclidean')
hclust_pca <- hclust(dist_mat_pca, method = 'ward.D')
```

```{r}
cut_pca <- cutree(hclust_pca, k=2)
plot(hclust_pca, main = "PCA Hierarchical Clustering Dendogram")
rect.hclust(hclust_pca,k=2, border = 2:6)
abline(h=80, col = 'red')

cluster.size.hc.pca <- table(cut_pca)
cluster.size.hc.pca
```


```{r}

#OS_MONTHS indicates the number of months from time of diagnosis to time of death or last follow up
clinical_cleaned <- clinicalData[which(clinicalData$PATIENT_ID %in% unique.patients.full.data) ,]

survival_DF <- data.frame(deceased = clinical_cleaned$OS_STATUS == "1:DECEASED")

#Create a column of months to death from diagnosis
indices <- which(clinical_cleaned$OS_STATUS == "1:DECEASED")

survival_DF$days_to_death = rep(NA, length(unique.patients.full.data))

for(i in indices) {
  survival_DF$days_to_death[i] = clinical_cleaned$OS_MONTHS[i]
}

```

```{r}
#SURVIVAL ANALYSIS on 2 PCA clusters

survival_DF$days_last_followup = clinical_cleaned$DAYS_LAST_FOLLOWUP / (365/12)

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to progression free disease for patients who
# are still alive
survival_DF$overall_survival = ifelse(survival_DF$deceased,
                                   survival_DF$days_to_death,
                                   survival_DF$days_last_followup)
#Assign the bmiClass vector containing results to the clinical dataframe

#Shows first 10 samples
head(survival_DF)

#Create a vector containing cluster groups for simplification
survival_DF$cluster_groups <- cut_pca

#Now that the survival time has been tagged with the censoring, we can add the categorical independent variable  `cluster groups`, and effectively create a formula

Surv(survival_DF$overall_survival, survival_DF$deceased) ~ survival_DF$cluster_groups

fit = survfit(Surv(overall_survival, deceased) ~ cluster_groups, data=survival_DF)

library("survival")
library("survminer")
library(ggplot2)

ggsurvplot(fit, data=survival_DF, pval=T)

```

```{r}
#SURVIVAL ANALYSIS on 2 clusters WITHOUT PCA *HERE

survival_DF$days_last_followup = clinical_cleaned$DAYS_LAST_FOLLOWUP/(365/12)

# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to progression free disease for patients who
# are still alive
survival_DF$overall_survival = ifelse(survival_DF$deceased,
                                   survival_DF$days_to_death,
                                   survival_DF$days_last_followup)
#Assign the bmiClass vector containing results to the clinical dataframe

#Shows first 10 samples
head(survival_DF)

#Create a vector containing cluster groups for simplification
survival_DF$cluster_groups <- cut_ward

#Now that the survival time has been tagged with the censoring, we can add the categorical independent variable  `cluster groups`, and effectively create a formula

Surv(survival_DF$overall_survival, survival_DF$deceased) ~ survival_DF$cluster_groups

fit = survfit(Surv(overall_survival, deceased) ~ cluster_groups, data=survival_DF)

library("survival")
library("survminer")
library(ggplot2)

ggsurvplot(fit, data=survival_DF, pval=T)

```















```{r}
cut_pca_5 <- cutree(hclust_pca, k=5)
plot(hclust_pca, main = "PCA Hierarchical Clustering Dendogram")
rect.hclust(hclust_pca,k=5, border = 2:6)
#abline(h=80, col = 'red')

cluster.size.5.pca <- table(cut_pca_5)
cluster.size.5.pca
```

```{r}
#SURVIVAL ANALYSIS

survival_DF$days_last_followup = clinical_cleaned$DAYS_LAST_FOLLOWUP/30.5


# create an "overall survival" variable that is equal to days_to_death
# for dead patients, and to progression free disease for patients who
# are still alive
survival_DF$overall_survival = ifelse(survival_DF$deceased,
                                   survival_DF$days_to_death,
                                   survival_DF$days_last_followup)
#Assign the bmiClass vector containing results to the clinical dataframe

#Shows first 10 samples
head(survival_DF)

#Create a vector containing cluster groups for simplification
survival_DF$cluster_groups <- cut_pca_5

#Now that the survival time has been tagged with the censoring, we can add the categorical independent variable  `cluster groups`, and effectively create a formula

Surv(survival_DF$overall_survival, survival_DF$deceased) ~ survival_DF$cluster_groups

fit = survfit(Surv(overall_survival, deceased) ~ cluster_groups, data=survival_DF)

library("survival")
library("survminer")
library(ggplot2)
ggsurvplot(fit, data=survival_DF)
```


```{r}
clust.sub.data <- data.frame(Patient = clinical_cleaned$PATIENT_ID, 
                             Subtype = clinical_cleaned$SUBTYPE, 
                             Clusters = cut_pca_5)

library(RColorBrewer)

ggplot(clust.sub.data, aes(x = cut_pca_5, fill = Subtype)) +
  geom_bar(position = "dodge") +
  labs(title = "Bar Plot of Subtypes within Clusters",
       x = "Cluster",
       y = "Count") +
  scale_fill_brewer(palette = "Set1", name = "Subtype") +
  theme_minimal() +  # Customize the theme if needed
  theme(legend.position = "top")
```
```{r}
clust.sub.data <- data.frame(Patient = clinical_cleaned$PATIENT_ID, 
                             Subtype = clinical_cleaned$SUBTYPE, 
                             Clusters = cut_pca)

library(RColorBrewer)

ggplot(clust.sub.data, aes(x = cut_pca, fill = Subtype)) +
  geom_bar(position = "dodge") +
  labs(title = "Bar Plot of Subtypes within Clusters",
       x = "Cluster",
       y = "Count") +
  scale_fill_brewer(palette = "Set1", name = "Subtype") +
  theme_minimal() +  # Customize the theme if needed
  theme(legend.position = "top")

```
